keps/sig-node/2400-node-swap/README.md

```diff
@@ -536,20 +611,22 @@ Here are specific improvements to be made:
   swap limit for workloads.
 - Investigate eviction behavior with swap enabled.
 
-
-#### Beta
-
-- Add support for controlling swap consumption at the pod level [via cgroups].
-- Handle usage of swap during container restart boundaries for writes to tmpfs
-  (which may require pod cgroup change beyond what container runtime will do at
-  container cgroup boundary).
+#### Beta 1
+- Enable Swap Support using Burstable QoS Pods only. 
 - Add swap memory to the Kubelet stats api.
 - Determine a set of metrics for node QoS in order to evaluate the performance
   of nodes with and without swap enabled.
-- Better understand relationship of swap with memory QoS in cgroup v2
```

Comment:
For cgroup v2, `memory.swap.high` may be helpful at the node level. We can set it in the node level using a factor like the other KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos for memory.

Reply:
Yeah, makes sense to align with v2 QoS work.

Reply:
I have added a section on using `memory.swap.high` - https://github.com/kubernetes/enhancements/pull/3957/files#diff-32d9cbd2b6817965ecee751dfe43b0d91269aee738a5799c9c9a68d04a329ec6R194

Reply:
I don't think that using swap throttling is the right approach here.

There are significant differences between `memory.high` and `memory.swap.high` and they are used in a completely different manner.

`memory.high` is described as `"the main mechanism to control memory usage."` and `"when hit, it throttles allocations by forcing them into direct reclaim to work off the excess"` [1].

However, `memory.swap.high` is described as [1]:
```
Swap usage throttle limit.  If a cgroup's swap usage exceeds
this limit, all its further allocations will be throttled to
allow userspace to implement custom out-of-memory procedures.

marks a point of no return for the cgroup. It is NOT
designed to manage the amount of swapping a workload does
during regular operation. Compare to memory.swap.max, which
prohibits swapping past a set amount, but lets the cgroup
continue unimpeded as long as other memory can be reclaimed.

Healthy workloads are not expected to reach this limit.
```

In addition, the Linux kernel commit message that brought `memory.swap.high` says [2] `That is not to say that the knob itself is equivalent to memory.high` and that `Slowing misbehaving tasks down gradually allows user space oom killers or other protection mechanisms to react`.

-----

Since we don't implement any user-space OOM killers, IIUC reaching the memory.swap.high limit would freeze the processes in the cgroup forever.

This makes sense, as swap memory isn't being reclaimed as often as ram, and it's much more expensive to do so. If processes in a cgroup would reach the `memory.swap.max` boundary, however, they would have to keep allocated memory in-ram. If `memory.high` is defined, as the KEP suggests, the processes would reach a point where they freeze up, forced to reclaim memory to continue their operation.

The bottom line is that processes that are out of memory **should reclaim ram memory, not swap memory**.

[1] https://www.kernel.org/doc/Documentation/admin-guide/cgroup-v2.rst
[2] https://lore.kernel.org/linux-mm/20200602044952.g4tpmtOiL%25akpm@linux-foundation.org/

Reply:
`Slowing misbehaving tasks down gradually allows user space oom killers or other protection mechanisms to react` 

IMO is a good thing. We have seen so many issues in the past where the workload would end up consuming the memory so fast that kernel won't get adequate time to react. In this scenario, the entire system would just freeze and the node would eventually transition into NOT_READY state in k8s.